cli/util/upload_bin.go

```diff
@@ -233,7 +233,10 @@ func fetchLatestMissingBlockIndex(ctx context.Context, p *pool.Pool, containerID
 				continue
 			}
 			if emptyBatchFound || (batch == numBatches && i == len(results)-1) {
-				return results[i].startIndex + searchBatchSize, nil
+				if batch < numBatches {
+					return results[i].endIndex, nil
+				}
+				return results[i].startIndex, nil
```

Comment:
it wont be if we have only 1 searcher

Reply:
I don't see the difference for the case of 1 searcher. Explain, please.

Reply:
Chain block height: 6689860
fetchLatestMissingBlockIndex
batch 668
results len 1
startIndex 6680000 endIndex 6690000  len(oids) 14742
emptyBatchFound false
batch, numBatches 668 668
i, len(results) 0 1

So its goes to `(batch == numBatches && i == len(results)-1)` and will `return return results[i].endIndex, nil` (6690000), but it should be startIndex as we know that 6690000 do not exist and we have "full" batch only because of duplicates.
so the result with 
```
if batch < numBatches {
					return results[i].endIndex, nil
				}
				return results[i].startIndex, nil
```
will be:
First block of latest incomplete batch uploaded to NeoFS container: 6680000.

Reply:
another case if we do expectedIndexCount := currentUploadedHeight / indexFileSize -1

 the searchers>1 will check 667, 668,669, 670 ...
we will still have this situation:
startIndex 6680000 endIndex 6690000 len(oids) 14742 (the previous one will be empty)
so it will return endIndex, which is 6690000

Reply:
maybe we should do expectedIndexCount := currentHeight / indexFileSize -maxParallelSearches
but then its not expectedIndexCount but something else.

Reply:
maybe we should do numBatches      = currentHeight / searchBatchSize -maxParallelSearches

Reply:
>  and we have "full" batch only because of duplicates.

Then this is the real problem that must be fixed, `fetchLatestMissingBlockIndex` must distinguish duplicates, otherwise the whole uploader just doesn't work correctly and may easily miss the subsequent not-finished batch. And actually the code that you provided doesn't fixes this issue completely. Consider the following case: uploaded only half of all blocks, the script is restarted. First `N` SEARCH requests return zero OIDs, hence `if results[i].numOIDs < searchBatchSize` check gives us `emptyBatchFound = true`. Then first half-empty batch is found, but due to duplicates the `N+1`-th SEARCH result returns more than `searchBatchSize` objects. Then this batch will be considered as completed and index of the first block of *next* batch will be returned as a result.

Checking `Index` attributes is not an option for us here, it's too expensive. Hence I propose a tiny hack: distinguish *empty` batches with exactly `0` number of elements from half-filled and full batches. Change the meaning of  `emptyBatchFound` to reflect if `0`-length batch is found. Then treat *all* batches that precedes the `0`-length one as non-empty.

> the searchers>1 will check 667, 668,669, 670

A note about this: the point is that checks are going in backwards direction, hence the second case that you've provided is just impossible.

Reply:
> Checking Index attributes is not an option for us here, it's too expensive. Hence I propose a tiny hack: distinguish *emptybatches with exactly0number of elements from half-filled and full batches. Change the meaning of emptyBatchFoundto reflect if0-length batch is found. Then treat *all* batches that precedes the 0`-length one as non-empty.

So if 0 len is found and the previous one is non-zero, we should return the startIndex of the non-zero batch.

> A note about this: the point is that checks are going in backwards direction, hence the second case that you've provided is just impossible.

Yes check of results -backward, but the search is forward. 
We start from the current height-> search for batches ( :current height+n] -> check the results of these searches in a backward manner. 

```
(base) ekaterinapavlova@MacBook-Air-4 neo-go % ./bin/neo-go util upload-bin --cid ANMHF1LcEhf4LsFQRCafst5PdHN1YTvhaZYk1EhLKMPH --wallet-config ./wallet-config.yml --block-attribute Block --index-attribute Index --rpc-endpoint https://rpc.morph.fs.neo.org --fs-rpc-endpoint apollo1.morphbits.io:8080 --fs-rpc-endpoint apollo2.morphbits.io:8080 --fs-rpc-endpoint apollo3.morphbits.io:8080 --timeout 3m --skip-blocks-uploading --index-file-size 10000

Chain block height: 6689717
fetchLatestMissingBlockIndex
batch 668
results len 20
6870000 6880000 0
6860000 6870000 0
6850000 6860000 0
6840000 6850000 0
6830000 6840000 0
6820000 6830000 0
6810000 6820000 0
6800000 6810000 0
6790000 6800000 0
6780000 6790000 0
6770000 6780000 0
6760000 6770000 0
6750000 6760000 0
6740000 6750000 0
6730000 6740000 0
6720000 6730000 0
6710000 6720000 0
6700000 6710000 0
6690000 6700000 0
emptyBatchFound true
batch, numBatches 668 668
i, len(results) 0 20
First block of latest incomplete batch uploaded to NeoFS container: 6680000
Uploading index files...
```

Reply:
The problem of duplicates refs. https://github.com/nspcc-dev/neo-go/issues/3652

Reply:
> But I think we can lose holes because searching does not align with uploading batches

It is aligned in fact. `results` contains search results for 10K-batches. Hence I don't think we'll miss gaps with the proposed adjustment.